src/uharfbuzz/_harfbuzz.pyx

```diff
@@ -1057,3 +1057,103 @@ cdef class DrawFuncs:
         self._close_path_func = func
         hb_draw_funcs_set_close_path_func(
             self._hb_drawfuncs, _close_path_func, <void*>user_data, NULL)
+
+cdef void free_allocated (hb_object_t* obj_list, unsigned int idx):
+    cdef unsigned int i
+    for i in range(idx):
+        if obj_list[i].real_links != NULL:
+            free(obj_list[i].real_links)
+        if obj_list[i].virtual_links != NULL:
+            free(obj_list[i].virtual_links)
+
+def hb_repack(subtables: List[bytes],
+              graphnodes: List[Tuple[List[Tuple[int, int, int]],
+                                     List[Tuple[int, int, int]]
+                                     ]]) -> bytes:
+    if len(subtables) != len(graphnodes):
+        return bytes(b'')
+    cdef:
+        unsigned int num_nodes = len(subtables)
+        unsigned int length = 0, prev_length = 0
+        dict idx_len_dict = {}
+    for i in range(num_nodes):
+        length += len(subtables[i])
+        idx_len_dict[i] = (prev_length, length)
+        prev_length = length
+
+    cdef bytes table_bytes = b''.join(subtables)
+    cdef char* table_data = table_bytes
+
+    cdef hb_object_t* obj_list = <hb_object_t*>calloc(num_nodes, sizeof(hb_object_t))
+    if obj_list == NULL:
+      return bytes(b'')
+
+    cdef:
+        unsigned int head, tail
+        unsigned int table_idx
+        tuple node
+        unsigned int real_link_num, virtual_link_num
+        hb_link_t* p = NULL
+    for i in range(num_nodes):
+        table_idx = num_nodes - i -1
+        head = (idx_len_dict.get(table_idx))[0]
+        tail = (idx_len_dict.get(table_idx))[1]
+
+        obj_list[i].head = table_data + head
+        obj_list[i].tail = table_data + tail
+
+        node = graphnodes[i]
+        #real_links
+        real_link_num = len(node[0])
+        obj_list[i].num_real_links = real_link_num
+        if real_link_num == 0:
+            obj_list[i].real_links = NULL
+        else:
+            p = <hb_link_t*>calloc(real_link_num, sizeof(hb_link_t))
```

Comment:
This code for creating the link list (and the similar one for virtual links) could be extracted into a separate function to improve readability.

Reply:
I tried to create a separate function but failed to build because:
This separate function needs to assign values to real_links member in hb_object_t struct with values coming from List[Tuple[int, int,int]]

for def function: i cannot pass in hb_object_t* type
for cdef function: I cannot pass in List[Tuple[int,int,int]]

Reply:
Oh, I think I might be able to do this by defining a class which is a wrapper of hb_object_t*, I'll work on this